File: ip-messaging-demo-android/src/main/java/com/twilio/ipmessaging/demo/MessageViewHolder.java
package com.twilio.ipmessaging.demo;

import java.util.Map;
import org.json.JSONObject;

import com.twilio.ipmessaging.Member;
import com.twilio.ipmessaging.Message;
import com.twilio.ipmessaging.demo.R;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Color;

import android.util.Base64;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;

import uk.co.ribot.easyadapter.ItemViewHolder;
import uk.co.ribot.easyadapter.PositionInfo;
import uk.co.ribot.easyadapter.annotations.LayoutId;
import uk.co.ribot.easyadapter.annotations.ViewId;

@LayoutId(R.layout.message_item_layout)
public class MessageViewHolder extends ItemViewHolder<MessageActivity.MessageItem>
{
    private static int[] HORIZON_COLORS = {
        Color.RED, Color.BLUE, Color.CYAN, Color.GREEN, Color.MAGENTA
    };

    @ViewId(R.id.body)
    TextView body;

    @ViewId(R.id.txtInfo)
    TextView txtInfo;

    @ViewId(R.id.singleMessageContainer)
    LinearLayout singleMessageContainer;

    @ViewId(R.id.consumptionHorizon)
    LinearLayout consumptionHorizon;

    @ViewId(R.id.avatar_message_left)
    ImageView imageViewLeft;

    @ViewId(R.id.avatar_message_right)
    ImageView imageViewRight;

    View view;

    public MessageViewHolder(View view)
    {
        super(view);
        this.view = view;
    }

    @Override
    public void onSetListeners()
    {
        view.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v)
            {
                OnMessageClickListener listener = getListener(OnMessageClickListener.class);
                if (listener != null) {
                    listener.onMessageClicked(getItem());
                }
            }
        });
    }

    @Override
    public void onSetValues(MessageActivity.MessageItem message, PositionInfo pos)
    {
        StringBuffer textInfo = new StringBuffer();
        if (message != null) {
            textInfo.append(message.getMessage().getAuthor());
            String dateString = message.getMessage().getTimeStamp();
            if (dateString != null) {
                textInfo.append(":").append(dateString);
            }
            txtInfo.setText(textInfo.toString());
            body.setText(message.getMessage().getMessageBody());

            boolean left =
                (message.getMessage().getAuthor().equals(message.getCurrentUser())) ? true : false;
            body.setBackgroundResource(left ? R.drawable.bubble_b : R.drawable.bubble_a);
            singleMessageContainer.setGravity(left ? Gravity.LEFT : Gravity.RIGHT);
            txtInfo.setGravity(left ? Gravity.LEFT : Gravity.RIGHT);
            consumptionHorizon.removeAllViews();
            imageViewLeft.setVisibility(View.GONE);
            imageViewRight.setVisibility(View.GONE);

            if (message.getMembers() != null && message.getMembers().getMembers() != null) {
                for (Member member : message.getMembers().getMembers()) {
                    if (message.getMessage().getAuthor().equals(
                            member.getUserInfo().getIdentity())) {
                        if (left)
                            fillUserAvatar(imageViewLeft, member);
                        else
                            fillUserAvatar(imageViewRight, member);
                    }
                    if (member.getLastConsumedMessageIndex() != null
                        && member.getLastConsumedMessageIndex()
                               == message.getMessage().getMessageIndex()) {
                        drawConsumptionHorizon(left, member);
                    }
                }
            }
        }
    }

    private void drawConsumptionHorizon(boolean left, Member member)
    {
        LinearLayout seenByView = new LinearLayout(getContext());
        seenByView.setOrientation(LinearLayout.VERTICAL);
        TextView identity = new TextView(getContext());
        identity.setGravity(left ? Gravity.LEFT : Gravity.RIGHT);
        identity.setLayoutParams(new LinearLayout.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));

        identity.setText(member.getUserInfo().getIdentity());
        View line = new View(getContext());
        line.setLayoutParams(new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, 5));
        line.setBackgroundColor(getMemberRgb(member.getUserInfo().getIdentity()));
        identity.setTextColor(getMemberRgb(member.getUserInfo().getIdentity()));
        identity.setTextSize(8);
        seenByView.addView(identity);
        seenByView.addView(line);
        consumptionHorizon.addView(seenByView);
    }

    private void fillUserAvatar(ImageView avatarView, Member member)
    {
        JSONObject attributes = member.getUserInfo().getAttributes();
        String     avatar = (String)attributes.opt("avatar");
        avatarView.setVisibility(View.VISIBLE);
        avatarView.setImageBitmap(null);
        if (avatar != null) {
            byte[] data = Base64.decode(avatar, Base64.NO_WRAP);
            Bitmap bitmap = BitmapFactory.decodeByteArray(data, 0, data.length);
            avatarView.setImageBitmap(bitmap);
        }
    }

    public interface OnMessageClickListener {
        void onMessageClicked(MessageActivity.MessageItem message);
    }

    public int getMemberRgb(String identity)
    {
        return HORIZON_COLORS[Math.abs(identity.hashCode()) % HORIZON_COLORS.length];
    }
}

File: chat-demo-android/src/main/kotlin/com/twilio/conversations/demo/views/MessageViewHolder.kt
package com.twilio.conversations.demo.views

import android.content.Context
import com.twilio.conversations.CallbackListener
import com.twilio.conversations.Participant
import com.twilio.conversations.User
import com.twilio.conversations.demo.R
import com.twilio.conversations.demo.TwilioApplication
import com.twilio.conversations.demo.activities.MessageActivity
import android.graphics.BitmapFactory
import android.graphics.Color
import android.net.Uri
import android.util.Base64
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.RelativeLayout
import android.widget.TextView
import kotterknife.bindView
import eu.inloop.simplerecycleradapter.SettableViewHolder
import java.io.File

class MessageViewHolder(val context: Context, parent: ViewGroup)
    : SettableViewHolder<MessageActivity.MessageItem>(context, R.layout.message_item_layout, parent)
{
    val avatarView: ImageView by bindView(R.id.avatar)
    val reachabilityView: ImageView by bindView(R.id.reachability)
    val body: TextView by bindView(R.id.body)
    val mediaView: ImageView by bindView(R.id.mediaPreview)
    val author: TextView by bindView(R.id.author)
    val date: TextView by bindView(R.id.date)
    val identities: RelativeLayout by bindView(R.id.consumptionHorizonIdentities)
    val lines: LinearLayout by bindView(R.id.consumptionHorizonLines)

    override fun setData(message: MessageActivity.MessageItem) {
        val msg = message.message

        author.text = msg.author
        body.text = msg.messageBody
        date.text = msg.dateCreated

        identities.removeAllViews()
        lines.removeAllViews()

//        for (member in message.membersList) {
//            if (msg.author.contentEquals(member.identity)) {
//                fillUserAvatar(avatarView, member)
//                fillUserReachability(reachabilityView, member)
//            }
//
//            if (member.lastConsumedMessageIndex != null && member.lastConsumedMessageIndex == message.message.messageIndex) {
//                drawConsumptionHorizon(member)
//            }
//        }

        if (msg.hasMedia()) {
            body.visibility = View.GONE
            mediaView.visibility = View.VISIBLE
            mediaView.setImageURI(Uri.fromFile(File(context.cacheDir, msg.mediaSid)))
        }
    }

    fun toggleDateVisibility() {
        date.visibility = if (date.visibility == View.GONE) View.VISIBLE else View.GONE
    }

    private fun drawConsumptionHorizon(member: Participant) {
        val ident = member.identity
        val color = getMemberRgb(ident)

        val identity = TextView(itemView.context)
        identity.text = ident
        identity.textSize = 8f
        identity.setTextColor(color)

        // Layout
        val params = RelativeLayout.LayoutParams(RelativeLayout.LayoutParams.MATCH_PARENT, RelativeLayout.LayoutParams.WRAP_CONTENT)
        val cc = identities.childCount
        if (cc > 0) {
            params.addRule(RelativeLayout.RIGHT_OF, identities.getChildAt(cc - 1).id)
        }
        identity.layoutParams = params

        val line = View(itemView.context)
        line.layoutParams = LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, 5)
        line.setBackgroundColor(color)

        identities.addView(identity)
        lines.addView(line)
    }

    private fun fillUserAvatar(avatarView: ImageView, member: Participant) {
        TwilioApplication.instance.basicClient.conversationsClient?.getAndSubscribeUser(member.identity, object : CallbackListener<User> {
            override fun onSuccess(user: User) {
                val attributes = user.attributes
                val avatar = attributes.jsonObject?.opt("avatar") as String?
                if (avatar != null) {
                    val data = Base64.decode(avatar, Base64.NO_WRAP)
                    val bitmap = BitmapFactory.decodeByteArray(data, 0, data.size)
                    avatarView.setImageBitmap(bitmap)
                } else {
                    avatarView.setImageResource(R.drawable.avatar2)
                }
            }
        })
    }

    private fun fillUserReachability(reachabilityView: ImageView, member: Participant) {
        if (!TwilioApplication.instance.basicClient.conversationsClient?.isReachabilityEnabled!!) {
            reachabilityView.setImageResource(R.drawable.reachability_disabled)
        } else {
            member.getAndSubscribeUser(object : CallbackListener<User> {
                override fun onSuccess(user: User) {
                    if (user.isOnline) {
                        reachabilityView.setImageResource(R.drawable.reachability_online)
                    } else if (user.isNotifiable) {
                        reachabilityView.setImageResource(R.drawable.reachability_notifiable)
                    } else {
                        reachabilityView.setImageResource(R.drawable.reachability_offline)
                    }
                }
            })
        }
    }

    fun getMemberRgb(identity: String): Int {
        return HORIZON_COLORS[Math.abs(identity.hashCode()) % HORIZON_COLORS.size]
    }

    companion object {
        private val HORIZON_COLORS = intArrayOf(Color.GRAY, Color.RED, Color.BLUE, Color.GREEN, Color.MAGENTA)
    }
}
